db/view/view.cc

```diff
@@ -2714,5 +2714,23 @@ void delete_ghost_rows_visitor::accept_new_row(const clustering_key& ck, const q
     }
 }
 
+std::chrono::microseconds calculate_view_update_throttling_delay(db::view::update_backlog backlog,
+                                                                 db::timeout_clock::time_point timeout) {
+    constexpr auto delay_limit_us = 1000000;
+    auto adjust = [] (float x) { return x * x * x; };
```

Comment:
Please explain this

Reply:
This is pre-existing code, which I guess @wmitros moved in his patch. As I explained in my document https://www.scylladb.com/2018/12/04/worry-free-ingestion-flow-control/, the controller can use **any** increasing function to convert the queue size to delay time. In this code, x queue size relative to some "maximum" (e.g., 10% of memory), so 0.1 means 1% of memory and 0.9 means 9% of memory. The reason why `x^3` was chosen (by Duarte) was to allow the queue to fill up a lot before taking drastic actions. For example, if the queue is half-full, x=0.5, the delay becomes 0.5^3, i.e., 0.125 of the maximum. Only when the queue reaches close to 1 do we reach close to the maximum delay.

We never actually proved in any way or even demonstrated that this function, `x^3`, is better than any other function in any way. In fact, the more we test this flow-control feature, the more we realize its implementation was completely broken. If you remember, a month after we released it, I discovered the delay was multiplied by zero (!) so had no effect. @wmitros discovered a few days ago that for small mutations the memory accounting was way off, so instead of reaching close to 10% of memory we could reach close to 80% of memory before the delay becomes noticably long. If we didn't see these problems in tests, it's not surprising that we couldn't tell if the choice of the function `x^3` was good or not.

Reply:
I'm resolving this conversation. This is pre-existing code moved from one source file to another - and it's good to move view-specific code from the kitchen-sink storage_proxy.cc to the right place, view.cc.